lightly/api/api_workflow_predictions.py

```diff
@@ -0,0 +1,152 @@
+                All singletons must be of the same type.
+
+            prediction_uuid_timestamp:
+                This timestamp is used as a key to distinguish different predictions for the same sample.
+                Get it e.g. via 'int(time.time())'.
+
+            progress_bar:
+                Tqdm progress bar to show how many prediction files have already been
+                uploaded.
+
+            max_workers:
+                Maximum number of workers uploading predictions in parallel.
+
+        Example:
+          >>> import time
+          >>> from tqdm import tqdm
+          >>> from lightly.api import ApiWorkflowClient
+          >>> from lightly.openapi_generated.swagger_client import (
+          >>>     PredictionTaskSchema,
+          >>>     TaskType,
+          >>>     PredictionTaskSchemaCategory,
+          >>> )
+          >>> from lightly.api.prediction_singletons import PredictionSingletonClassificationRepr
+          >>>
+          >>> client = ApiWorkflowClient(
+          >>>     token="MY_LIGHTLY_TOKEN", dataset_id="MY_DATASET_ID"
+          >>> )
+          >>>
+          >>> timestamp = int(time.time())
+          >>> filenames = client.get_filenames()
+          >>> filename_to_prediction_singletons_dummy = {
+          >>>     filename: [PredictionSingletonClassificationRepr(taskName="my-task", categoryId=i%4, score=0.9, probabilities=[0.1, 0.2, 0.3, 0.4])]
+          >>>     for i, filename in enumerate(filenames)
+          >>> }
+          >>> client.create_or_update_predictions(
+          >>>     filename_to_prediction_singletons_dummy,
+          >>>     prediction_uuid_timestamp=timestamp,
+          >>>     progress_bar=tqdm(desc="Uploading predictions", total=len(filenames), unit=" predictions")
+          >>> )
+
+
+        """
+
+        samples = retry(
+            self._samples_api.get_samples_partial_by_dataset_id,
+            dataset_id=self.dataset_id,
+            mode=SamplePartialMode.FILENAMES,
+        )
+        api_filename_to_sample_id = {sample.file_name: sample.id for sample in samples}
+
+        if any(filename_to_upload not in api_filename_to_sample_id for filename_to_upload in filename_to_prediction_singletons.keys()):
+            raise ValueError(f"Some filenames to upload are not found as samples on the server.")
+
+        def upload_prediction(filename_and_predictions_tuple) -> bool:
+            (filename, predictions) = filename_and_predictions_tuple
+            sample_id = api_filename_to_sample_id[filename]
+            prediction_singletons_for_sending = [vars(singleton) for singleton in predictions]
+            retry(
+                self._predictions_api.create_or_update_prediction_by_sample_id,
+                body=prediction_singletons_for_sending,
+                dataset_id=self.dataset_id,
+                sample_id=sample_id,
+                prediction_uuid_timestamp=prediction_uuid_timestamp,
+            )
```

Comment:
Minor, optional: This function can be moved outside of the enclosing function. It will need an extra argument and then needs to be passed in as a partial function:
```py
def _upload_prediction(prediction_uuid_timestamp, filename_and_predictions_tuple) -> bool:
    ...

def create_or_update_predictions(...):
    ...

    upload_prediction_partial = functools.partial(_upload_prediction, prediction_uuid_timestamp)
    for _ in executor.map(upload_prediction_partial, filename_to_prediction_singletons.items()):
        ...
```

Reply:
It will need even more extra arguments: the `self` and the `api_filename_to_sample_id`. Thus I think that extracting it makes the code more ugly.